app/src/main/java/ru/gorshenev/themesstyles/data/repositories/chat/ChatRepository.kt

```diff
@@ -0,0 +1,178 @@
+    }
+
+    fun getQueueReactions(queueId: String, lastId: Int): Observable<GetEmojiEventsResponse> {
+        return api.getEmojiEventsFromQueue(queueId, lastId)
+    }
+
+    fun getMessage(messageId: Int) = api.getMessage(messageId)
+
+
+    fun onEmojiClick(emojiName: String, messageId: Int): Observable<CreateReactionResponse> {
+        return api.getMessage(id = messageId, applyMarkdown = true)
+            .map { response ->
+                response.message.reactions.filter { reaction -> reaction.emojiName == emojiName }
+                    .any { it.userId == Reactions.MY_USER_ID }
+            }
+            .concatMapSingle { isMyClick -> updateEmoji(messageId, emojiName, isMyClick) }
+            .subscribeOn(Schedulers.io())
+    }
+
+    private fun updateEmoji(
+        msgId: Int,
+        emojiName: String,
+        isMyClick: Boolean = false
+    ): Single<CreateReactionResponse> {
+        return when {
+            isMyClick -> api.deleteEmoji(msgId = msgId, emojiName)
+            else -> api.addEmoji(msgId = msgId, emojiName)
+        }
+    }
+
+    fun sendMessage(message: String): Single<CreateMessageResponse> {
+        return api.sendMessage(
+            to = streamName,
+            topic = topicName,
+            content = message
+        )
+            .subscribeOn(Schedulers.io())
+    }
+
+    private fun updateDb(
+        messagesInDb: List<MessageEntity>, newMessage: MessageResponse, topicName: String
+    ): Completable {
+        val msgDoesNotExist = messagesInDb.all { it.msgId != newMessage.msgId }
+        val msgExist = messagesInDb.any { it.msgId == newMessage.msgId }
+
+        return when {
+            messagesInDb.isEmpty() || msgDoesNotExist && messagesInDb.size != 50 -> {
+                Completable.fromCallable {
+                    messageDao.insertMessageWithReactions(
+                        newMessage.toEntity(topicName),
+                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
+                    )
+                }
+            }
+
+            msgExist -> {
+                Completable.fromCallable {
+                    messageDao.updateMessageReactions(
+                        newMessage.msgId,
+                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
+                    )
+                }
+            }
+
+            msgDoesNotExist && messagesInDb.size == 50 -> {
+                Completable.fromCallable {
+                    messageDao.deleteFirstAndAddNewMessage(
+                        messagesInDb.first().msgId,
+                        newMessage.toEntity(topicName),
+                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
+                    )
+                }
+            }
+
+            else -> {
+                Log.d("database", "add to data ERROR")
+                Completable.error(Error("add to data ERROR"))
+            }
+        }
+    }
```

Comment:
private fun updateDb(
        messagesInDb: List<MessageEntity>, newMessage: MessageResponse, topicName: String,
    ): Completable {
        val msgExist = messagesInDb.any { it.msgId == newMessage.msgId }

        return when {
            msgExist -> {
                Completable.fromCallable {
                    messageDao.updateMessageReactions(
                        newMessage.msgId,
                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
                    )
                }
            }

            messagesInDb.size < 50 -> {
                Completable.fromCallable {
                    messageDao.insertMessageWithReactions(
                        newMessage.toEntity(topicName),
                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
                    )
                }
            }

            else -> {
                Completable.fromCallable {
                    messageDao.deleteFirstAndAddNewMessage(
                        messagesInDb.first().msgId,
                        newMessage.toEntity(topicName),
                        newMessage.reactions.toEntity(newMessage.msgId, topicName)
                    )
                }
            }
        }
    }

Reply:
+